Sketch for ICM20948 integration into MPU9250 driver.

flochir opened the pull request:
Hello again!
I sketched out an approach to adding support for ICM20948 to the MPU9250 driver as suggested (#9345, #8624). Here it is (far from functional, mind you).
If you agree that this is a good path to follow, I will continue with this.
The changes with the biggest impact on the original MPU9250 driver parts should be done. It builds and the driver seems to be running OK with MPU9250 on my Pixhawk 2.1.

Outline of the changes/thoughts:
The ICM20948 not only uses different register addresses, but spreads them over 4 banks of registers. So to access a particular register, one has to first select the containing bank by writing to the BANK_SEL register.
This would actually not be needed to only support the magnetometer (writes to default bank 0 suffice), but adding the gyro and accel and generally supporting the device needs this. I could just hack in magnetometer support, but I guess full support would be nice. Or is this superfluous?

So I started to add multiple device support analogous to MPU6000.

The additional information about the register banks I have added by expanding the register definitions for ICM20948 to 16 bit and using the lower two bits of the MSB. All the write/read functions take 16 bit register addresses anyway and use bit 0x8000 to encode bus speed.
I added a bit of code to the read and write functions which should make this work transparently for the rest of the code (just using 8/16 bit addresses).
I had also played with a lookup table for register/bank pairs but this keeps the definitions more in one place and is probably a bit more memory efficient.

Best, Flochir

flochir commented:
Great!
One thought I'd like to ask your opinion for. As this is built into the Here GPS module, which is usually mounted upon a GPS boom, I'm aiming for a switch to allow only the magnetometer to be enabled. I'm thinking about enabling all of it with no parameter, and only the magnetometer with an "-M" switch present.

My guess is that on a (more or less) shaking/vibrating boom the provided accel/gyro data would be noisy. Or do you think the PX4 sensor fusion would be able to pull advantage from the additional data nonetheless? Also maybe it makes sense to only disable the accelerometer and keep the gyro?

bkueng commented on src/drivers/imu/mpu9250/mpu9250.cpp:
```diff
@@ -995,11 +1047,26 @@ MPU9250::gyro_ioctl(struct file *filp, int cmd, unsigned long arg)
 
+int
+MPU9250::select_register_bank(uint8_t bank) {
+	if (bank != _selected_bank ) {
```
Shouldn't we set `_selected_bank` in here?

flochir replied:
Thanks for another catch!

bkueng commented:
@flochir I tested it with the Here GPS on a Pixhawk 2., and had to do this change:
```diff
diff --git a/src/drivers/imu/mpu9250/main.cpp b/src/drivers/imu/mpu9250/main.cpp
index 6f222730d9..cf7bcb502d 100644
--- a/src/drivers/imu/mpu9250/main.cpp
+++ b/src/drivers/imu/mpu9250/main.cpp
@@ -77,6 +77,9 @@
 #define MPU_DEVICE_PATH_ACCEL_EXT   "/dev/mpu9250_accel_ext"
 #define MPU_DEVICE_PATH_GYRO_EXT    "/dev/mpu9250_gyro_ext"
 #define MPU_DEVICE_PATH_MAG_EXT     "/dev/mpu9250_mag_ext"
+#define MPU_DEVICE_PATH_ACCEL_EXT2  "/dev/mpu9250_accel_ext2"
+#define MPU_DEVICE_PATH_GYRO_EXT2   "/dev/mpu9250_gyro_ext2"
+#define MPU_DEVICE_PATH_MAG_EXT2    "/dev/mpu9250_mag_ext2"

 /** driver 'main' command */
 extern "C" { __EXPORT int mpu9250_main(int argc, char *argv[]); }
@@ -118,7 +121,7 @@ struct mpu9250_bus_option {
 #  endif
 #  if defined(PX4_I2C_BUS_EXPANSION)
     { MPU9250_BUS_I2C_EXTERNAL, MPU_DEVICE_TYPE_MPU9250, MPU_DEVICE_PATH_ACCEL_EXT, MPU_DEVICE_PATH_GYRO_EXT, MPU_D
EVICE_PATH_MAG_EXT, &MPU9250_I2C_interface, false, PX4_I2C_BUS_EXPANSION, NULL },
-    { MPU9250_BUS_I2C_EXTERNAL, MPU_DEVICE_TYPE_ICM20948, MPU_DEVICE_PATH_ACCEL_EXT, MPU_DEVICE_PATH_GYRO_EXT, MPU_
DEVICE_PATH_MAG_EXT, &MPU9250_I2C_interface, false, PX4_I2C_BUS_EXPANSION, NULL },
+    { MPU9250_BUS_I2C_EXTERNAL, MPU_DEVICE_TYPE_ICM20948, MPU_DEVICE_PATH_ACCEL_EXT2, MPU_DEVICE_PATH_GYRO_EXT2, MP
U_DEVICE_PATH_MAG_EXT2, &MPU9250_I2C_interface, false, PX4_I2C_BUS_EXPANSION, NULL },
 #  endif
 #endif
 #ifdef PX4_SPIDEV_MPU
```

But `mpu9250 -X start` still fails with:
```
MPU9250_I2C on I2C bus 1 at 0x69 (bus: 100 KHz, max: 400 KHz)
ERROR [mpu9250] Reg 3 is:224 s/b:234 Tries:2
ERROR [mpu9250] Reg 512 is:234 s/b:0 Tries:2
ERROR [mpu9250] Reg 3 is:224 s/b:234 Tries:1
ERROR [mpu9250] Reg 512 is:234 s/b:0 Tries:1
ERROR [mpu9250] Reg 3 is:224 s/b:234 Tries:0
ERROR [mpu9250] Reg 512 is:234 s/b:0 Tries:0
ERROR [mpu9250] Exiting! Device failed to take initialization
```
Do you have it working already?

flochir commented on src/drivers/imu/mpu9250/mpu9250.h:
```diff
@@ -193,10 +200,61 @@
 /**
  * Report conversation within the mpu, including command byte and
  * interrupt status.
+#define ADDR_ID_ICM         0x00
```
No, thanks for the catch :)

flochir commented on src/drivers/imu/mpu9250/mpu9250.h:
```diff
@@ -341,10 +402,18 @@ class MPU9250 : public device::CDev
 	// reset
 #define MPU9250_NUM_CHECKED_REGISTERS 11
-	static const uint8_t	_checked_registers[MPU9250_NUM_CHECKED_REGISTERS];
-	uint8_t			_checked_values[MPU9250_NUM_CHECKED_REGISTERS];
```
Sure, good suggestion!
Just so I get this right - that would safe some memory, but what would change for the type of allocation? Also, would it matter? The driver object is created once at system start?

flochir replied:
Ah - no need to explain anymore ;)

flochir commented:
This is not yet working. Already on it though (and integrated your suggestions on my end, will update soon).
This pull request was just a sketch to discuss the basic concept of integrating ICM20948 support before I continue. If you need the Here GPS magnetometer working quickly, I commited an almost working standalone driver in #9345. It reads sensor data from the magnetometer, but has a few remaining issues. After Lorenz Meiers suggestion to better integrate it with the MPU9250 driver I concentrated on this. It will take a few more days though, as I have some other stuff in parallel right now.

Thanks for the review, I'll update soon with a recent version!

flochir commented:
Here's a bit of progress on this. Any thoughts welcome.

bkueng commented:
> It will take a few more days though, as I have some other stuff in parallel right now.

Sure, no problem. I missed your initial comment that it's not functional yet.

LorenzMeier commented:
@flochir Where is this standing?

flochir commented:
@LorenzMeier 
Had a bit of delay for the last weeks because of some high priority deadlines, but I'm on it again.
I've integrated everything and get data, but there's some quirks left I'm debugging (lots of bad transfers). Will update soon (I hope ;)

flochir commented:
@LorenzMeier @bkueng 
I agree to push for the 1.8 release. Can you give me a timeline on this? I still have stability issues with the driver and wasn't able to put as much time into this as I would have liked. This should finally be better next week.
I'll upload my working version until tomorrow.

flochir commented:
I pushed my recent version and would appreciate any hints. Maybe I'm missing something obvious. Reading data from the sensors works, but it keeps crashing the PX4. Loading the driver and taking a single reads seems to work fine, but it crashes fast when the data is published.
The version I pushed delivers a measurement when you call "mpu9250 -X start; mpu9250 -X info" in nsh. If the orb publishing / poll notification for the gyro data is uncommented in measure() (line 1746 and on in mpu9250.cpp), it crashes very fast. 

How would you go about debugging this?

Another issue is that there are registers checked bad quite often and bad transfers. Could be the I2C bus, but it would be quite noisy then - is this the case on the pixhawk, in your experience? I need to get to the bottom of this, too.

I noticed you pushed some changes to the mpu9250 driver on the master branch but didn't merge them yet, will do as soon as possible.

LorenzMeier commented on src/drivers/imu/mpu9250/mpu9250.cpp:
```diff
@@ -995,11 +1178,51 @@ MPU9250::gyro_ioctl(struct file *filp, int cmd, unsigned long arg)
+	if (ret != OK) {
+		return ret;
+	}
+	up_udelay(100);
```
This stalls the complete CPU. Don't do this in a driver in general and ONLY as exception during setup. This is critical to resolve.

flochir replied:
Thanks - replaced up_udelay() with usleep().

LorenzMeier commented on src/drivers/imu/mpu9250/mag.cpp:
```diff
@@ -586,40 +607,54 @@ MPU9250_mag::ak8963_setup(void)
-		_parent->modify_reg(MPUREG_USER_CTRL, 0, BIT_I2C_MST_RST);
-		up_udelay(100);
+		_parent->modify_reg(AK_MPU_OR_ICM(MPUREG_USER_CTRL, ICMREG_20948_USER_CTRL), 0, BIT_I2C_MST_RST);
+		usleep(100);
```
usleep doesn't work in interrupt context - in which context is this running?

flochir replied:
This should not  be called in interrupt context - it's run on driver start and in the event of a reset during ioctl or in an error event during polling. I'll check if I can take some of those out however.

LorenzMeier replied:
I would change that back to up_udelay() here.

LorenzMeier commented on src/drivers/imu/mpu9250/mag.cpp:
```diff
@@ -586,40 +607,54 @@ MPU9250_mag::ak8963_setup(void)
-			ak8963_setup_master_i2c();
-			write_reg(AK8963REG_CNTL2, AK8963_RESET);
+				_parent->modify_reg(AK_MPU_OR_ICM(MPUREG_USER_CTRL, ICMREG_20948_USER_CTRL), 0, BIT_I2C_MST_RST);
+				usleep(100);
```
Same question here.

flochir replied:
Same here.

LorenzMeier replied:
Would change this to up_udelay()

LorenzMeier commented on src/drivers/imu/mpu9250/mpu9250.cpp:
```diff
@@ -995,11 +1178,51 @@ MPU9250::gyro_ioctl(struct file *filp, int cmd, unsigned long arg)
+		PX4_WARN("BANK retries: %d", count);
+
+
+		usleep(20);
```
This doesn't work due to the scheduling quantum (minimum of one millisecond) and doesn't work in interrupt context. Furthermore 20 microseconds is quite low, just the switch from the write() to the read() call will take the CPU this time frame. Have you tried just leaving this out entirely?

Both udelay and usleep are problematic in operation. The way to do it is to re-schedule the driver to re-run, not to delay the whole system or thread.

flochir replied:
These are not very important as the success of switching the register bank is checked and re-tried. Tried to minimize checks and retries (with resulting I2C communication) by using delays here. I'll just take those out and re-think the other delays.These are not very important as the success of switching the register bank is checked and re-tried. Tried to minimize checks and retries (with resulting I2C communication) by using delays here. I'll just take those out and re-think the other delays.

LorenzMeier commented:
@flochir PX4 drivers follow the task queue design pattern. Every task needs to be self-contained and minimise execution time. When you need to do two things that are time-wise separated, you need to schedule two runs, like explained here: https://en.wikipedia.org/wiki/Scheduling_(computing)#task_queue

flochir commented on src/drivers/imu/mpu9250/mag.cpp:
```diff
@@ -442,8 +456,8 @@ MPU9250_mag::passthrough_read(uint8_t reg, uint8_t *buf, uint8_t size)
 {
 	set_passthrough(reg, size);
 	usleep(25 + 25 * size); // wait for the value to be read from slave
```
If the use of usleep() and similar is an issue, this may also be something to be reviewed (in passthrough_read() as well)

DanielePettenuzzo commented on src/drivers/imu/mpu9250/mpu9250.cpp:
```diff
@@ -368,7 +397,7 @@ MPU9250::init()
 
 	/* do CDev init for the mag device node, keep it optional */
-	if (_whoami == MPU_WHOAMI_9250) {
+	if ( _whoami == MPU_WHOAMI_9250 || _device_type == MPU_DEVICE_TYPE_ICM20948 ) {
```
shouldn't we check _whoami also for the ICM20948 instead of _device_type == MPU_DEVICE_TYPE_ICM20948?

flochir replied:
Yes, this was a bit provisional to get going and no lasting solution. I cleaned it up and introduced a new device type for the MPU6500, will push later.

DanielePettenuzzo commented on src/drivers/imu/mpu9250/mpu9250.h:
```diff
@@ -193,6 +200,135 @@
+ * interrupt status.
+ */
+struct ICMReport {
+	uint8_t		accel_x[2];
```
MPU9250_I2C::read uses the MPUReport structure therefore this could cause issues in sending the ICMReport structure. I think we at least need a uint8_t cmd to receive something while we transfer the cmd byte in the i2c transfer. I will also further investigate this.

DanielePettenuzzo commented on src/drivers/imu/mpu9250/mpu9250.cpp:
```diff
@@ -532,44 +584,87 @@ int MPU9250::reset_mpu()
+		case MPU_WHOAMI_6500:
+			_num_checked_registers = MPU9250_NUM_CHECKED_REGISTERS;
+			_checked_registers = _mpu9250_checked_registers;
+			memset(_checked_values, 0, sizeof(_mpu9250_checked_registers));
```
eg.
memset(_checked_values, 0, sizeof(_checked_values));

flochir replied:
Thanks a lot! It had to be something stupid, seems I went blind there... :-/

flochir commented on src/drivers/imu/mpu9250/main.cpp:
```diff
@@ -119,21 +124,24 @@ struct mpu9250_bus_option {
+	{ MPU9250_BUS_SPI_INTERNAL, MPU_DEVICE_TYPE_MPU9250, MPU_DEVICE_PATH_ACCEL, MPU_DEVICE_PATH_GYRO, MPU_DEVICE_PATH_MAG, &MPU9250_SPI_interface, true, PX4_SPI_BUS_SENSORS, PX4_SPIDEV_MPU, NULL },
 #endif
 #ifdef PX4_SPIDEV_MPU2
 	{ MPU9250_BUS_SPI_INTERNAL2, MPU_DEVICE_PATH_ACCEL_1, MPU_DEVICE_PATH_GYRO_1, MPU_DEVICE_PATH_MAG_1, &MPU9250_SPI_interface, true, PX4_SPI_BUS_SENSORS, PX4_SPIDEV_MPU2, NULL },
```
Thanks, fixed!

flochir commented:
Updated and fixed the issues pointed out by DanielePettenuzzo:
* Fixed the memset problem, the hard fault in probe() is fixed
* Added an own device type for MPU6500 and added it to the bus options list. I used the same device paths as for MPU9250 for now (as it doesn't seem to be intended to be used in parallel with mpu9250), at the moment this gives the error "diver start failed" when the driver tries to take the already used device path for MPU6500. Is adding a dedicated set of device paths the right thing to do here?

There are hard faults now when _mag->measure() is called in mpu9250.cpp on line 1618.
Also I now get this error when starting with ICM20948:
```
nsh> mpu9250 -X start
MPU9250_I2C on I2C bus 1 at 0x69 (bus: 100 KHz, max: 400 KHz)
INFO  [mpu9250] accel cutoff set to 30.00 Hz
INFO  [mpu9250] gyro cutoff set to 80.00 Hz
AK8963_I2C on I2C bus 1 at 0x0c (bus: 100 KHz, max: 400 KHz)
nsh> ls
nsh: ls: out of memory
```
Will further debug this, any hints very welcome.

flochir commented on src/drivers/imu/mpu9250/mpu9250.cpp:
```diff
@@ -1456,8 +1481,15 @@ MPU9250::check_registers(void)
+		        // reset_mpu();
+		    }
+		    else {
+			    write_reg(MPUREG_PWR_MGMT_1, BIT_H_RESET);
```
I doubt if this is working for the MPU9250. It just resets the chip, but doesn't set it up again?

flochir commented on src/drivers/imu/mpu9250/mpu9250.cpp:
```diff
@@ -494,17 +494,17 @@ int MPU9250::reset_mpu()
+		/*
+		 * ICM20948 needs a bit of time here, else register settings will be garbled.
+		 */
+	    usleep(200);
```
@LorenzMeier 
Without these delays, sensor initialization is getting unstable. I'll check again later to rule out other causes, when things are more stable and/or minimize them. Would you actually stop the task and reschedule during initialization to avoid this?

bkueng commented:
>  Is adding a dedicated set of device paths the right thing to do here?

I think so.

> nsh: ls: out of memory

The board is out of memory. For testing, can you disable something else from the `rcS` startup script, such as `gps start` or `navigator start`? Always make sure that you have enough free RAM for testing, you can check with `free`.

flochir commented:
It seems to have memory left:
```
NuttShell (NSH)
nsh> ls
/:
 dev/
 etc/
 fs/
 obj/
 proc/
nsh> free
             total       used       free    largest
Mem:        235856     186864      48992      48592
nsh> mpu9250 -X start
MPU9250_I2C on I2C bus 1 at 0x69 (bus: 100 KHz, max: 400 KHz)
INFO  [mpu9250] accel cutoff set to 30.00 Hz
INFO  [mpu9250] gyro cutoff set to 80.00 Hz
AK8963_I2C on I2C bus 1 at 0x0c (bus: 100 KHz, max: 400 KHz)
nsh> ls
nsh: ls: out of memory
nsh> free
             total       used       free    largest
Mem:        235856     191184      44672      41408
nsh>
```

LorenzMeier commented:
@flochir That suggests some sort of software fault, e.g. due to excessive buffer allocations.

LorenzMeier commented on src/drivers/imu/mpu9250/mag.cpp:
```diff
@@ -586,40 +607,54 @@ MPU9250_mag::ak8963_setup(void)
-		retries = 10;
+	/* No sensitivity adjustments available for AK09916/ICM20948 */
+	if(_parent->_device_type == MPU_DEVICE_TYPE_MPU9250) {
+		if (retries > 0) {
```
Putting this here into the check for 9250 can change the global data flow.

LorenzMeier commented on src/drivers/imu/mpu9250/main.cpp:
```diff
@@ -119,21 +124,29 @@ struct mpu9250_bus_option {
-	{ MPU9250_BUS_I2C_INTERNAL, MPU_DEVICE_PATH_ACCEL, MPU_DEVICE_PATH_GYRO, MPU_DEVICE_PATH_MAG,  &MPU9250_I2C_interface, false, PX4_I2C_BUS_ONBOARD, PX4_I2C_OBDEV_MPU9250, NULL },
+#  if defined(PX4_I2C_BUS_ONBOARD) && defined(PX4_I2C_OBDEV_MPU9250)
+	{ MPU9250_BUS_I2C_INTERNAL, MPU_DEVICE_TYPE_MPU9250, MPU_DEVICE_PATH_ACCEL, MPU_DEVICE_PATH_GYRO, MPU_DEVICE_PATH_MAG,  &MPU9250_I2C_interface, false, PX4_I2C_BUS_ONBOARD, PX4_I2C_OBDEV_MPU9250, NULL },
+    { MPU9250_BUS_I2C_INTERNAL, MPU_DEVICE_TYPE_MPU6500, MPU_DEVICE_PATH_ACCEL, MPU_DEVICE_PATH_GYRO, MPU_DEVICE_PATH_MAG,  &MPU9250_I2C_interface, false, PX4_I2C_BUS_ONBOARD, PX4_I2C_OBDEV_MPU9250, NULL },
```
Why is device type 6500 here?

flochir replied:
MPU6500 was not differentiated from MPU9250 except by its whoami value, which was used to decide if the magnetometer was activated or not.
I was unsure if the MPU9250 could actually be used in parallel with MPU9250. But as the driver is going to support yet one more device I think its the cleanest solution to give every device supported by the driver its own device type, probe it at setup and only use the device type to differentiate the used device later on.

LorenzMeier commented on src/drivers/imu/mpu9250/mpu9250.cpp:
```diff
@@ -110,14 +112,36 @@ const uint8_t MPU9250::_checked_registers[MPU9250_NUM_CHECKED_REGISTERS] = { MPU
 									   };
 
 
+const uint16_t MPU9250::_icm20948_checked_registers[ICM20948_NUM_CHECKED_REGISTERS] = {
```
Why was this added?

flochir replied:
The ICM20948 uses different register addresses for the same/similar functions as the MPU9250, so another set of checked registers is needed. Also the ICM distributes its registers accross 4 register banks, so every register is 1 byte address + 2 bit information for its bank. I put both into the now 16bit address value as the write_reg/write_checked_reg functions use 16 bit integer addresses already to encode the bus speed, and the register bank value now goes alongside.

LorenzMeier commented on src/drivers/imu/mpu9250/mpu9250.cpp:
```diff
@@ -457,24 +484,48 @@ int MPU9250::reset()
+		/*
+		 * ICM20948 needs a bit of time here, else register settings will be garbled.
+		 */
+	    usleep(200);
```
Try up_udelay()

davids5 commented:
>Outline of the changes/thoughts:
The ICM20948 not only uses different register addresses, but spreads them over 4 banks of registers. So to access a particular register, one has to first select the containing bank by writing to the BANK_SEL register.
This would actually not be needed to only support the magnetometer (writes to default bank 0 suffice), but adding the gyro and accel and generally supporting the device needs this. I could just hack in magnetometer support, but I guess full support would be nice. Or is this superfluous?

Based on the comment about the register set I would like to understand if this sensor should have it's own driver.

LorenzMeier commented:
@davids5 Could you please review this? If it turns out to be a too hard effort we might want to resort to that, yes.

https://github.com/PX4/Firmware/pull/9345

DanielePettenuzzo commented on src/drivers/imu/mpu9250/mpu9250.cpp:
```diff
@@ -457,24 +484,48 @@ int MPU9250::reset()
+		 */
+	    usleep(200);
 
-	usleep(1000);
```
@flochir I tried putting this delay back and at least I can get the mpu9250 to run now. Obviously to test it I uncommented the gyro integrator and uorb publishing that you commented out for debugging in the measure function.

DanielePettenuzzo replied:
if this is necessary as it seems, we can reschedule the reset_mpu using the work queue

flochir replied:
Thanks! Can you try again with my second latest commit (8481016873e4b640801a8502c9de9807b41c0a4c, the latest doesn't publish sensor data)? The mpu9250 is working for me with this one, but with only the two short delays as up_udelay().
Rescheduling probably still makes sense later on, this could be called after setup if the sensor or the bus have problems and a reset is triggered.

flochir commented:
I don't think the integration of ICM20948 itself is the problem anymore. I just pushed a version with all the publishing/system notifications commented out, see the output below. The sensor reads sensical values when tested (turning/accelerating it produces believable output).
I'm having trouble getting it stable though, it crashes fast as soon as I publish gyrometer or magnetometer reports, while with accelerometer reporting alone it works.
The registers checked bad (see output below) seem to be connected to problems with the register bank selection (sometimes registers get read from the wrong bank because the selection seems to fail or the register bank is switched between selection and read, don't know yet).
This would need to be solved for a standalone driver too.
For the MPU9250 the driver is working for me. With the sensor communication and the crashes for the ICM part fixed it should work.

```
nsh> mpu9250 -X info
state @ 20024f50
Device type:20948
mpu9250_read: 41736 events, 193828040us elapsed, 4644us avg, min 478us max 12990us 756.130us rms
mpu9250_acc_read: 0 events
mpu9250_gyro_read: 0 events
mpu9250_bad_trans: 11291 events
mpu9250_bad_reg: 45 events
mpu9250_good_trans: 22494 events
mpu9250_reset: 0 events
mpu9250_dupe: 8479 events
accel queue	2/144 (2/1 @ 20025620)
gyro queue	2/144 (2/2 @ 20025720)
mag queue	2/112 (2/2 @ 20025920)
checked_next: 1
reg 0003:00 was bad 11
reg 0006:01 was bad 00
reg 0007:00 was bad 01
reg 000f:12 was bad 00
reg 0010:00 was bad 12
reg 0011:01 was bad 04
reg 0012:00 was bad 01
reg 0200:ea should be 04
reg 0200:ea was bad 00
reg 0201:07 was bad 00
reg 0211:04 was bad 00
reg 0214:00 should be 07
reg 0214:00 was bad 04
reg 0215:11 was bad 07
temperature: 60.0
accel cutoff set to      30.00 Hz
gyro cutoff set to      80.00 Hz
 sensor_mag_s
	timestamp: 280545552 (0.004668 seconds ago)
	error_count: 0
	x: 0.186
	y: 0.435
	z: -0.275
	range_ga: 48.000
	scaling: 0.002
	temperature: 59.953
	device_id: 289033
	x_raw: 104
	y_raw: 291
	z_raw: -185
	is_external: 0
 sensor_accel_s
	timestamp: 280471919 (0.094444 seconds ago)
	integral_dt: 7526
	error_count: 11336
	x: -9.033
	y: -1.703
	z: 1.320
	x_integral: -0.061
	y_integral: -0.006
	z_integral: 0.024
	temperature: 59.953
	range_m_s2: 156.906
	scaling: 0.005
	device_id: 1468681
	x_raw: -1627
	y_raw: -127
	z_raw: 745
	temperature_raw: 9008
 sensor_gyro_s
	timestamp: 280471919 (0.107468 seconds ago)
	integral_dt: 18125373657369678868
	error_count: 11336
	x: -0.102
	y: -1.706
	z: 0.214
	x_integral: 0.000
	y_integral: 0.000
	z_integral: 0.000
	temperature: 59.953
	range_rad_s: 34.907
	scaling: 0.001
	device_id: 2386185
	x_raw: -60
	y_raw: -808
	z_raw: 48
	temperature_raw: 9008
```

flochir commented:
With register checking disabled, the bad transfers vanish and the sensor is running fine about 50% of the starts. Once it's running one or two seconds, it's working stable.
If it fails however, it crashes with a hard fault after a few hundred to thousand measurements the other times.
But with sensor data publishing disabled it's working stable and keeps measuring.

I tried to debug the hard fault and get these from R8:
(gdb) info line *0x08006df1
Line 152 of "mm_heap/mm_sem.c" starts at address 0x8006df0 <mm_takesemaphore+8> and ends at 0x8006df6 <mm_takesemaphore+14>.
(gdb) info line *0x080178cc
Line 84 of "mm_heap/mm_free.c" starts at address 0x80178cc <mm_free+14> and ends at 0x80178d8 <mm_free+26>.

Which is mm_takesemaphore() called by mm_free(). I'll try to get on from there. Any hints how to best extract more information?

LorenzMeier commented:
Which registers are you checking? These transfers can be very expensive runtime-wise and might be a real concern.

flochir commented:
I'm checking all registers set during reset_mpu(). I tried to do it analogous to what the driver does with the original mpu9250. The checked registers are:

ICMREG_20948_USER_CTRL,
ICMREG_20948_PWR_MGMT_1,
ICMREG_20948_PWR_MGMT_2,
ICMREG_20948_INT_PIN_CFG,
ICMREG_20948_INT_ENABLE,
ICMREG_20948_INT_ENABLE_1,
ICMREG_20948_INT_ENABLE_2,
ICMREG_20948_INT_ENABLE_3,
ICMREG_20948_GYRO_SMPLRT_DIV,
ICMREG_20948_GYRO_CONFIG_1,
ICMREG_20948_GYRO_CONFIG_2,
ICMREG_20948_ACCEL_SMPLRT_DIV_1,
ICMREG_20948_ACCEL_SMPLRT_DIV_2,
ICMREG_20948_ACCEL_CONFIG,
ICMREG_20948_ACCEL_CONFIG_2

flochir commented on src/drivers/imu/mpu9250/mpu9250.cpp:
```diff
@@ -498,7 +503,7 @@ int MPU9250::reset_mpu()
 		 * ICM20948 needs a bit of time here, else register settings will be garbled.
 		 */
-	    up_udelay(200);
+	    up_udelay(2000);
```
The reset needs time here, the flag will sometimes not be cleared even after 2ms. I will reschedule this later to get rid of the hard delay. But for easier debugging I will keep it for now - it does not seem a likely reason for the instability in itself and only called once during setup.

davids5 commented:
>With register checking disabled, the bad transfers vanish and the sensor is running fine about 50% of the starts. Once it's running one or two seconds, it's working stable.
If it fails however, it crashes with a hard fault after a few hundred to thousand measurements the other times.
But with sensor data publishing disabled it's working stable and keeps measuring.

>I tried to debug the hard fault and get these from R8:
(gdb) info line *0x08006df1
Line 152 of "mm_heap/mm_sem.c" starts at address 0x8006df0 <mm_takesemaphore+8> and ends at 0x8006df6 <mm_takesemaphore+14>.
(gdb) info line *0x080178cc
Line 84 of "mm_heap/mm_free.c" starts at address 0x80178cc <mm_free+14> and ends at 0x80178d8 <mm_free+26>.

>Which is mm_takesemaphore() called by mm_free(). I'll try to get on from there. Any hints how to best extract more information?

Is the code allocating memory in an ISR (hrt callback)?

flochir commented:
With I2C, the hrt should not be used for scheduling - _use_hrt is set to false, and the scheduling is done through the HP work queue:

```
void
MPU9250::start()
{
	/* make sure we are stopped first */
	stop();

	/* discard any stale data in the buffers */
	_accel_reports->flush();
	_gyro_reports->flush();
	_mag->_mag_reports->flush();

	if (_use_hrt) {
		/* start polling at the specified rate */
		hrt_call_every(&_call,
			       1000,
			       _call_interval - MPU9250_TIMER_REDUCTION,
			       (hrt_callout)&MPU9250::measure_trampoline, this);

	} else {
#ifdef USE_I2C
		/* schedule a cycle to start things */
		work_queue(HPWORK, &_work, (worker_t)&MPU9250::cycle_trampoline, this, 1);
#endif
	}

}
```
HRT timestamps are used however.

flochir commented:
Here's the corresponding part of the MPU9250 initializer list:
```
#if defined(USE_I2C)
	_work {},
	_use_hrt(false),
#else
	_use_hrt(true),
#endif
```
Also, I've checked that the code actually enters the work queue path.

flochir commented:
I tried to find some system behind the hard faults / out of memory occurences. In my last test run, I had 6 hard faults and 3 occurences of "out of memory" executing ls after starting the driver.
The hard faults seem to happen after a random time, an extreme one took about 1-2 seconds to happen.

The hard faults yielded the following R8 value classes. Does this give any clue?

(gdb) info line *0x08006df1
Line 152 of "mm_heap/mm_sem.c" starts at address 0x8006df0 <mm_takesemaphore+8>
   and ends at 0x8006df6 <mm_takesemaphore+14>.
(gdb) info line *0x080178e2
Line 89 of "mm_heap/mm_free.c" starts at address 0x80178e2 <mm_free+36>
   and ends at 0x80178e8 <mm_free+42>.

(gdb) info line *0x08006df1
Line 152 of "mm_heap/mm_sem.c" starts at address 0x8006df0 <mm_takesemaphore+8>
   and ends at 0x8006df6 <mm_takesemaphore+14>.
(gdb) info line *0x08017918
Line 123 of "mm_heap/mm_free.c" starts at address 0x8017918 <mm_free+90>
   and ends at 0x801791e <mm_free+96>.

(gdb) info line *0x08006df1
Line 152 of "mm_heap/mm_sem.c" starts at address 0x8006df0 <mm_takesemaphore+8>
   and ends at 0x8006df6 <mm_takesemaphore+14>.
(gdb) info line *0x080178e8
Line 105 of "mm_heap/mm_free.c" starts at address 0x80178e8 <mm_free+42>
   and ends at 0x80178ec <mm_free+46>.

(gdb) info line *0x08006df1
Line 152 of "mm_heap/mm_sem.c" starts at address 0x8006df0 <mm_takesemaphore+8>
   and ends at 0x8006df6 <mm_takesemaphore+14>.
(gdb) info line *0x08017906
Line 114 of "mm_heap/mm_free.c" starts at address 0x8017906 <mm_free+72>
   and ends at 0x8017912 <mm_free+84>.

davids5 commented:
@flochir - This looks like a memory corruption issue. Theses are hard to find, as the clues can be very misleading. Causes can be freeing memory that was not allocated from the heap, memory overwrites, stack under flow.  Have a look at stack usage increase based on your changes. Did you ups the stack size?

flochir commented:
@davids5:
Didn't try that yet!
I raised  CONFIG_SCHED_HPWORKSTACKSIZE in platforms/nuttx/nuttx-configs/px4fmu-v2/nsh from 1800 to 2200, but still get hard faults. Is this what you had in mind, and the right place to do it? I'm compiling for px4fmu-v3 (Pixhawk 2.1 Cube).

davids5 commented:
@flochir 

If you have the debug serial console connected. There should be a stack penetration figures dumped when the hardfault happens. They are also available in 'top' 

CONFIG_SCHED_HPWORKSTACKSIZE Would be one place. What if the corruptions happens in the setup when the app is launched?

See mpu9250/CMakeLists.txt

px4_add_module(
	MODULE drivers__mpu9250
	MAIN mpu9250
	STACK_MAIN 1200
	COMPILE_FLAGS
	SRCS
		mpu9250.cpp
		mpu9250_i2c.cpp
		mpu9250_spi.cpp
		main.cpp
		gyro.cpp
		mag.cpp
		mag_i2c.cpp
	DEPENDS
	)

flochir commented:
@davids5:
Seems like that did the trick!!
I raised STACK_MAIN to 1800, and just did 30 reboots and driver loads without a problem. Thank you very much! I wasn't aware of that setting.

I'll test and clean up the code, try to fix the register checking and add a magnetometer only flag. I'd also squash my commits into a new pull request, or is merging the whole bugfixing commit history desirable?

flochir commented:
@davids5, @LorenzMeier 
Based on the discussion above, what are your thoughts about this integration into the MPU9250 driver vs. a separate ICM20948 driver? If you think it's the better solution, it would not be hard to do at this point.
The registers are more or less completely different, and the register banks go on top of that. Also a separate driver may be more towards the plans for a new unified driver base/base class?
However the control flow and general architecture is very similar, and the magnetometer is nearly identical in both devices. The driver consumes more memory and flash now, but a bit less then having two different drivers loaded I would gess (which would be the case e.g. for a Pixhawk 2.1 cube with a Here GPS attached).

davids5 commented:
@flochir - Great!

We should understand why the stack growth 600 bytes.

Squashing to cogent commits is best. 

It is a hard call to split or not to split. Memory is less ONLY if only one driver is needed in the build. The coupling works to expand the testing of all the device types on all buses.

flochir commented:
@davids5 
The stack consumption probably didn't grow that much. I just raised it a fair bit to quickly see if it makes a difference.

flochir commented:
Finally had a quiet time to get on with this.
Just pushed a (still preliminary) version that's working for both mpu9250 and icm20948 on Pixhawk 2.1 with Here GPS as an external compass.
I also tested the function of the whole IMU including accel and gyro, disabling all other IMUs in rc.sensors I was able to calibrate the Pixhawk with just the ICM20948.
No further tests yet, just got it running and would appreciate any feedback.

Register checking is still disabled, will look into that one. Is there a way to make two I2C transfers "atomic", so that no other call is done in between? So that a register select and a subsequent read/write don't get interrupted by another register select.

mhkabir commented:
Closing this as continued in #10116
